walkeeper/src/timeline.rs

```diff
@@ -197,8 +206,44 @@ impl Timeline {
         }
     }
 
-    fn _stop_wal_senders(&self) {
-        self.notify_wal_senders(END_REPLICATION_MARKER);
+    // Notify WAL senders that it's time to stop sending WAL
+    pub fn stop_streaming(&self) {
+        let mut shared_state = self.mutex.lock().unwrap();
+        // Ensure that safekeeper sends WAL up to the last known committed LSN.
+        // It guarantees that pageserver will receive all the latest data
+        // before walservice disconnects.
+        shared_state.stop_lsn = Some(shared_state.notified_commit_lsn);
+        trace!(
+            "Stopping WAL senders. stop_lsn: {}",
+            shared_state.sk.flush_lsn
+        );
+        self.cond.notify_all();
+    }
+
+    // Reset stop_lsn notification,
+    // so that WAL senders will continue sending WAL
+    pub fn continue_streaming(&self) {
+        let mut shared_state = self.mutex.lock().unwrap();
+        shared_state.stop_lsn = None;
+    }
+
+    // Check if it's time to stop streaming to the given replica
+    pub fn check_stop_streaming(&self, replica_id: usize) -> Result<()> {
+        let shared_state = self.mutex.lock().unwrap();
+
+        // If stop_lsn is set, it's time to shutdown streaming.
+        if let Some(stop_lsn_request) = shared_state.stop_lsn {
+            let replica_state = shared_state.replicas[replica_id].unwrap();
+            // Lsn::MAX means that we don't know the latest LSN yet.
+            // XXX How should we handle this case?
```

Comment:
Seems like it's fine, we just should wait until it becomes known -- and ensure KeepAlives are actually sent regularly, even if nothing happens -- which is true as I see.

Why not use Option here instead of ::MAX?

Also I haven't checked whether stop happens if commit_lsn is 0, in the very beginning.

Reply:
> Why not use Option here instead of ::MAX?

I used it in the same manner as ReplicaState.disk_consistent_lsn.
Maybe it's worth refactoring both, though.

> Also I haven't checked whether stop happens if commit_lsn is 0, in the very beginning.

Hmm. Could you describe this scenario in more detail?
I guess it wouldn't stop because if safekeeper hasn't streamed anything it means that it hasn't received the feedback as well.

Reply:
Well, commit_lsn becomes non-zero only after majority of safekeepers acked WAL receival for the first time (and safekeeper hears about that); it might not happen before the disconnection. So maybe better handle that separately.

Reply:
Fixed